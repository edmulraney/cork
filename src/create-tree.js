const elementRegex = /(<[^>]+>)/g
const getElements = str => str.split(elementRegex)
const isTextNode = element => element !== null && element.indexOf('<') === -1
const isCloseElement = element => element.indexOf('</') !== -1
const isOpenElement = element => element.indexOf('>') !== -1 && element.indexOf('/>') === -1
const isSelfCloseElement = element => element !== null && element.indexOf('/>') !== -1
const isOpenComponent = element => element !== null && /(<[A-Z])/.test(element) && element.indexOf('/>') === -1

const getElementName = str => {
  const spaceIndex = str.indexOf(' ')
  const hasSpace = spaceIndex !== -1
  return str.replace(/[><\/]/g, '').substring(0, hasSpace ? spaceIndex - 1 : str.length)
}
const getProps = element => {
  const attributes = element.substring(element.indexOf(' ') + 1, element.lastIndexOf('"') + 1)
  const propsArray = attributes.split(/\="([^"]+)"/)
  propsArray.pop() // remove empty last item
  const props = {}
  propsArray.forEach((prop, index) => (index % 2 === 0) && (props[prop.trim()] = getProp(propsArray[index + 1])))
  return props
}
const createNode = (element, Components) => {
  const props = isTextNode(element) ? ({ nodeValue: element })  : getProps(element)
  if (isOpenComponent(element)) {
    return Components[getElementName(element)](props)
  }

  const node = {
    debug: element,
    type: isTextNode(element) ? 'text' : getElementName(element),
    props: {
      children: [],
      ...props,
    }
  }
  return node
}
export const createTree = (html, Components) => {
  const elements = getElements(html)
  const rootElement = { type: 'root', props: { children: [] } }
  const stack = [rootElement]
  let elementIndex = 0
  
  while (elementIndex !== elements.length) {
    let parent = stack[stack.length - 1]
    const element = elements[elementIndex++]
    const isEmptyNode = element.trim() === ''
    if (isEmptyNode) continue

    if (isCloseElement(element)) {
      stack.pop()
      parent = stack[stack.length - 1]
      continue
    }

    const node = createNode(element, Components)
    
    if (isSelfCloseElement(element) || isTextNode(element)) {
      parent.props.children.push(node)
      continue
    }
    
    if (isOpenElement(element)) {
      parent.props.children.push(node)
      stack.push(node)
      continue
    }
    
    
  }  
  console.log(JSON.stringify(stack[0], null, 2))
  return stack[0].props.children[0] // ignore root container
}

