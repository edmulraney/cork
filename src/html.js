const hasComponent = html => /(<[A-Z])/.test(html)
const html = (statics, ...dynamics) => {
  if (!hasComponent(statics.join(''))) {
    return {
      type: 'fragment',
      statics,
      dynamics,
      components: null,
    }
  }
  return (...components) => {
    return {
      type: 'fragment',
      statics,
      dynamics,
      components: components.reduce((xs, x) => ({...xs, [x.name]: x }), {})
    }
  }
}

export {
  html,
}