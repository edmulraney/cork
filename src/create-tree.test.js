import { html } from './html'
import { createTree } from './create-tree'
import * as mocks from './create-tree.test.expectations'

const debug = tree => console.log(JSON.stringify(tree, null, 2))
// TODO: make these snapshots instead so you dont have to store expectations

describe('createTree', () => {
  it('prevents single nested child fragments', () => {
    // we dont want devs to nest inline html declarations, they should be declared as one fragment
    const spec = html`
      <div>
        ${html`<h1>Hello</h1>`}
      </div>
    `
    const result = createTree(spec)
    expect(result).toMatchSnapshot()
  })

  it('handles correct ordering of nested children', () => {
    const spec = html`
      <ul>
        <b>${[1, 2, 3].map(x => html`<li>${x}</li>`)}</b>
        <h1>foo</h1>
      </ul>
    `

    const result = createTree(spec)
    expect(result).toMatchSnapshot()
  })

  it('handles child component', () => {
    const CustomH1 = props => html`<h1>${props.children}</h1>`
    const spec = html`
      <div>
        <CustomH1>foo</CustomH1>
      </div>`(CustomH1)
    const result = createTree(spec)
    expect(result).toMatchSnapshot()
  })

  it('handles child component array', () => {
    const CustomUl = props => html`<ul>${props.children}</ul>`
    const CustomLi = props => html`<li>${props.children}</li>`
    const spec = html`
      <CustomUl>
        <b>${[1, 2, 3].map(x => html`<CustomLi>${x}</CustomLi>`(CustomLi))}</b>
      </CustomUl>`(CustomUl)
    const result = createTree(spec)
    expect(result).toMatchSnapshot()
  })

  it('handles dynamic element props', () => {
    const onClick = e => onClick('foo')
    const spec = html`<button id="${1}" onclick="${onClick}">Foo and ${'bar'} :)</button>`
    const result = createTree(spec)
    expect(result).toMatchSnapshot()
  })

  it('handles static element props', () => {})

  it('handles static and dynamic element props', () => {})

  it('handles static component props', () => {})

  it('handles dynamic component props', () => {})

  it('handles static and dynamic component props', () => {})


})
