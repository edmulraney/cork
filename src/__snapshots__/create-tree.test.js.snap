// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTree handles child component 1`] = `
Object {
  "debug": "<div>",
  "props": Object {
    "children": Array [
      Object {
        "debug": "<CustomH1>",
        "props": Object {
          "children": Array [
            Object {
              "debug": "foo",
              "props": Object {
                "children": Array [],
                "nodeValue": "foo",
              },
              "type": "text",
            },
          ],
        },
        "type": [Function],
      },
    ],
  },
  "type": "div",
}
`;

exports[`createTree handles child component array 1`] = `
Object {
  "debug": "<CustomUl>",
  "props": Object {
    "children": Array [
      Object {
        "debug": "<b>",
        "props": Object {
          "children": Array [
            Object {
              "debug": "<CustomLi>",
              "props": Object {
                "children": Array [
                  Object {
                    "debug": "1",
                    "props": Object {
                      "children": Array [],
                      "nodeValue": "1",
                    },
                    "type": "text",
                  },
                ],
              },
              "type": [Function],
            },
            Object {
              "debug": "<CustomLi>",
              "props": Object {
                "children": Array [
                  Object {
                    "debug": "2",
                    "props": Object {
                      "children": Array [],
                      "nodeValue": "2",
                    },
                    "type": "text",
                  },
                ],
              },
              "type": [Function],
            },
            Object {
              "debug": "<CustomLi>",
              "props": Object {
                "children": Array [
                  Object {
                    "debug": "3",
                    "props": Object {
                      "children": Array [],
                      "nodeValue": "3",
                    },
                    "type": "text",
                  },
                ],
              },
              "type": [Function],
            },
          ],
        },
        "type": "b",
      },
    ],
  },
  "type": [Function],
}
`;

exports[`createTree handles correct ordering of nested children 1`] = `
Object {
  "debug": "<ul>",
  "props": Object {
    "children": Array [
      Object {
        "debug": "<b>",
        "props": Object {
          "children": Array [
            Object {
              "debug": "<li>",
              "props": Object {
                "children": Array [
                  Object {
                    "debug": "1",
                    "props": Object {
                      "children": Array [],
                      "nodeValue": "1",
                    },
                    "type": "text",
                  },
                ],
              },
              "type": "li",
            },
            Object {
              "debug": "<li>",
              "props": Object {
                "children": Array [
                  Object {
                    "debug": "2",
                    "props": Object {
                      "children": Array [],
                      "nodeValue": "2",
                    },
                    "type": "text",
                  },
                ],
              },
              "type": "li",
            },
            Object {
              "debug": "<li>",
              "props": Object {
                "children": Array [
                  Object {
                    "debug": "3",
                    "props": Object {
                      "children": Array [],
                      "nodeValue": "3",
                    },
                    "type": "text",
                  },
                ],
              },
              "type": "li",
            },
          ],
        },
        "type": "b",
      },
      Object {
        "debug": "<h1>",
        "props": Object {
          "children": Array [
            Object {
              "debug": "foo",
              "props": Object {
                "children": Array [],
                "nodeValue": "foo",
              },
              "type": "text",
            },
          ],
        },
        "type": "h1",
      },
    ],
  },
  "type": "ul",
}
`;

exports[`createTree handles dynamic element props 1`] = `
Object {
  "debug": "<button id=\\"",
  "props": Object {
    "children": Array [
      Object {
        "debug": "Foo and ",
        "props": Object {
          "children": Array [],
          "nodeValue": "Foo and ",
        },
        "type": "text",
      },
      Object {
        "debug": "bar",
        "props": Object {
          "children": Array [],
          "nodeValue": "bar",
        },
        "type": "text",
      },
      Object {
        "debug": " :)",
        "props": Object {
          "children": Array [],
          "nodeValue": " :)",
        },
        "type": "text",
      },
    ],
    "id": 1,
    "onclick": [Function],
  },
  "type": "button",
}
`;

exports[`createTree prevents single nested child fragments 1`] = `
Object {
  "debug": "<div>",
  "props": Object {
    "children": Array [],
  },
  "type": "div",
}
`;
